libraries/chain/include/eosio/chain/protocol_state_object.hpp

```diff
@@ -16,9 +16,14 @@ namespace eosio { namespace chain {
     */
    class protocol_state_object : public chainbase::object<protocol_state_object_type, protocol_state_object>
    {
-      OBJECT_CTOR(protocol_state_object, (activated_protocol_features)(preactivated_protocol_features)(whitelisted_intrinsics))
-
    public:
+      template<typename Constructor>
+      protocol_state_object(Constructor&& c, chainbase::constructor_tag) :
+         id(0),
+         whitelisted_intrinsics(*activated_protocol_features.get_allocator(this)) {
```

Comment:
I wonder if we should create a `shared_flat_multimap` that derives from `boost::interprocess::flat_multimap` that provides a default constructor that passes in `pinnable_mapped_file::get_allocator<char>(obj)`. Make every `shared_*` type have this functionality like `shared_cow_string` and `shared_cow_vector`.

Reply:
I'm not understanding the suggestion?

Reply:
What you have here is fine. I was thinking alternatively we could have:

```
template <typename K, typename V>
class shared_flat_multimap : public boost::interprocess::flat_multimap< K, V, std::less<K>, allocator< std::pair<K,V>>> {
public:
   shared_flat_multimap() : boost::interprocess::flat_multimap( pinnable_mapped_file::get_allocator<char>(this) ) {}
...
```

Reply:
As an alternative to what? `protocol_state_object`?

Reply:
As an alternative to: 
```
   template<typename K, typename V>
   using shared_flat_multimap = boost::interprocess::flat_multimap< K, V, std::less<K>, allocator< std::pair<K,V> > >;
```

Reply:
Oh, thanks, I finally understand your suggestion. Let me give that a try!

Reply:
I can't do that unfortunately. `pinnable_mapped_file::get_allocator<char>(this)` returns a `std::optional`, and when empty it means we should use the heap (std::allocator()). But `boost::interprocess::flat_multimap` has no easy way to support two different allocators. 

I guess I could do this by creating a `proxy` allocator which would use one or the other. I'm happy to do that, it may be overkill for just this one instance of using shared_flat_multimap, but if we ever used more it would be cleaner.

Reply:
I think it is fine as is.